lib/std/secrets.nim

```diff
@@ -0,0 +1,226 @@
+## | JS(Web Browser)| `getRandomValues`_|
+## | Other platforms| `/dev/urandom`_|
+##
+## .. _BCryptGenRandom: https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom
+## .. _getrandom: https://man7.org/linux/man-pages/man2/getrandom.2.html
+## .. _/dev/urandom: https://en.wikipedia.org/wiki//dev/random
+## .. _getentropy: https://www.unix.com/man-page/mojave/2/getentropy
+## .. _SecRandomCopyBytes: https://developer.apple.com/documentation/security/1399291-secrandomcopybytes?language=objc
+## .. _getentropy openbsd: https://man.openbsd.org/getentropy.2
+## .. _getrandom freebsd: https://www.freebsd.org/cgi/man.cgi?query=getrandom&manpath=FreeBSD+12.0-stable
+## .. _getRandomValues: https://www.w3.org/TR/WebCryptoAPI/#Crypto-method-getRandomValues
+## 
+
+import std/os
+
+runnableExamples:
+  doAssert urandom(0).len == 0
+  doAssert urandom(10).len == 10
+  doAssert urandom(20).len == 20
+  doAssert urandom(120).len == 120
+
+
+
+when defined(posix):
+  import std/posix
+
+  const batchSize = 256
+
+  template processReadBytes(readBytes: int, p: pointer, res: var int) =
+    if readBytes == 0:
+      break
+    elif readBytes > 0:
+      inc(res, readBytes)
+      cast[ptr pointer](p)[] = cast[pointer](cast[ByteAddress](p) + readBytes)
+    else:
+      if osLastError().int in {EINTR, EAGAIN}:
+        discard
+      else:
+        res = -1
+        break
+
+  proc getDevUrandom(p: var openArray[byte], size: Natural): int =
+    let size = p.len
+    if size == 0:
+      return
+
+    let fd = posix.open("/dev/urandom", O_RDONLY)
+
+    if fd > 0:
+      var stat: Stat
+      if fstat(fd, stat) != -1 and S_ISCHR(stat.st_mode):
+        let
+          chunks = (size - 1) div batchSize
+          left = size - chunks * batchSize
+
+        var base = 0
+        for i in 0 ..< chunks:
+          let readBytes = posix.read(fd, addr p[base], batchSize)
+          if readBytes < 0:
+            return readBytes
+          inc(base, batchSize)
+
+        result = posix.read(fd, addr p[base], left)
+      discard posix.close(fd)
+
+when defined(js):
+  import std/private/jsutils
+
+  proc getRandomValues(arr: Uint8Array) {.importjs: "window.crypto.getRandomValues(#)".}
+
+  const
+    maxBrowserCryptoBufferSize = 256
+
+  proc urandom*(p: var openArray[byte]): int =
+    let size = p.len
+    if size > 0:
+      let chunks = (size - 1) div maxBrowserCryptoBufferSize
+      var base = 0
+      for i in 0 ..< chunks:
+        for j in 0 ..< maxBrowserCryptoBufferSize:
+          var src = newUint8Array(maxBrowserCryptoBufferSize)
+          getRandomValues(src)
+          p[base + j] = src[j]
+
+        inc(base, maxBrowserCryptoBufferSize)
+
+      let left = size - chunks * maxBrowserCryptoBufferSize
+      var src = newUint8Array(left)
+      getRandomValues(src)
+      for i in 0 ..< left:
+        p[base + i] = src[i]
+
+elif defined(windows):
+  type
```

Comment:
optional, could be done in future work: `std/private/wintypes.nim` to avoid each module re-defining those? (later can be refactored to avoid duplication with winlean, io.nim and others)